File: backend/app/models/citizen_report.py
from datetime import datetime

from geoalchemy2 import Geometry
from sqlalchemy import DateTime, ForeignKey, String, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.db import Base


class CitizenReport(Base):
    """A real report submitted by a citizen through the public map's "Buat
    Laporan" page -- distinct from community-reports.geojson, which is the
    team's own 42 survey points shown as explicitly-labeled sample/placeholder
    data. Rows here are genuine, whatever a citizen actually submits.

    Tied to a specific surveyed halte (like MaintenanceTask's halte_id) --
    the report page has the citizen pick an existing halte on a map rather
    than drop an arbitrary point, so DISHUB's dashboard can show "Laporan
    Warga" grouped under the halte it's about, the same way it already does
    for field survey notes.
    """

    __tablename__ = "citizen_report"

    report_id: Mapped[str] = mapped_column(String, primary_key=True)
    halte_id: Mapped[str] = mapped_column(String, ForeignKey("halte_survey.halte_id"))

    # Plain text, not a foreign key to an account -- there is no login/account
    # system in this project (see task.py's assigned_to for the same reasoning).
    reporter_name: Mapped[str] = mapped_column(String)
    description: Mapped[str] = mapped_column(String)
    # Every photo the reporter attached, in the order they were picked:
    # [{"url": "/uploads/<uuid>.jpg", "annotated_url": "/uploads/<uuid>.annotated.jpg"|None}, ...].
    # THIS list is the authoritative record of what was submitted. The two
    # scalar columns below hold the FIRST entry and are kept only because a
    # number of readers were written against a single-photo report (the
    # dashboard's report card, the cleanup hook that reclaims orphaned files,
    # the approved-photo strip). Every write path sets the list and the
    # scalars together from the same save result, so the two cannot disagree;
    # any new reader should use `photos` and treat the scalars as "primary
    # photo" shorthand.
    photos: Mapped[list | None] = mapped_column(JSONB, nullable=True)

    # Relative paths under /uploads, e.g. "/uploads/<uuid>.jpg" -- None if the
    # reporter didn't attach that media type.
    photo_url: Mapped[str | None] = mapped_column(String, nullable=True)
    # The same photo with the detector's boxes drawn on it, rendered by the
    # detector itself (see photo_detection.render_annotated). This is what the
    # dashboard shows DISHUB -- a dispatcher should see what the model saw,
    # not have to reconstruct it from a class list. None when the photo
    # produced no detection or the render failed; photo_url is always kept, so
    # the raw evidence never depends on the detector being up.
    photo_annotated_url: Mapped[str | None] = mapped_column(String, nullable=True)
    video_url: Mapped[str | None] = mapped_column(String, nullable=True)
    # "baru" (not yet dispatched -- shown on the dashboard map as a marker
    # needing attention) -> "diproses" (a MaintenanceTask now exists for it,
    # see POST /citizen-reports/{id}/dispatch). There is no "selesai" status:
    # once the dispatched task reaches "selesai" the row is deleted outright
    # (see update_task_status in routers/task.py) rather than kept around in
    # a third state, per the dispatcher workflow this models.
    status: Mapped[str] = mapped_column(String, default="baru")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    # Kept alongside halte_id (rather than derived from the halte's own
    # location) so a report always records exactly where the citizen tapped,
    # even if that's a few meters off from the halte's surveyed point.
    geom: Mapped[str] = mapped_column(Geometry(geometry_type="POINT", srid=4326))

    # Raw output of the YOLO infrastructure detector on photo_url, plus the
    # facility-level reading derived from it. Stored per-report so the
    # original evidence behind any halte_survey value the detector filled in
    # stays auditable -- see app/services/photo_detection.py.
    # Shape: {"detections": [...], "observed": {...}, "model": "...", "error": str|None}
    ai_detections: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    ai_analyzed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

File: frontend/src/types/citizenReport.ts
export interface CitizenReportPhoto {
  url: string;
  // The detector's own render of `url` with its boxes drawn on it -- what
  // "Laporan Warga" shows DISHUB. Null when that photo produced no detection
  // or the render failed; `url` is always the fallback.
  annotated_url: string | null;
}

export interface CitizenReport {
  report_id: string;
  halte_id: string;
  reporter_name: string;
  lat: number;
  lon: number;
  description: string;
  photo_url: string | null;
  // The detector's own render of photo_url with its boxes drawn on it -- what
  // "Laporan Warga" shows DISHUB. Null when the photo produced no detection
  // (or the render failed); photo_url is always the fallback.
  photo_annotated_url: string | null;
  // Every photo on the report, in submission order; photo_url and
  // photo_annotated_url are the first entry. Use this list to render a
  // gallery -- reports submitted before multi-photo uploads existed come back
  // as a one-entry list, so there is no need to handle a missing list.
  photos: CitizenReportPhoto[];
  video_url: string | null;
  status: string;
  created_at: string;
  // Detector output for photo_url and which halte_survey facility variables
  // it filled in -- see backend/app/services/photo_detection.py. None/empty
  // when no photo was attached or the detector found nothing new.
  ai_detections: Record<string, unknown> | null;
  ai_analyzed_at: string | null;
  halte_updated: Record<string, string>;
}

export interface CitizenReportCreateInput {
  lat: number;
  lon: number;
  halteId: string;
  reporterName: string;
  description: string;
  photos?: File[];
  video?: File;
}
